hydrolib/core/basemodel.py

```diff
@@ -95,31 +96,57 @@ class FileModel(BaseModel, ABC):
     it doesn't error, but actually initializes the `FileModel`.
     """
 
+    __slots__ = ["__weakref__"]
+    # Use WeakValueDictionary to keep track of file paths with their respective parsed file models.
+    _file_models_cache: WeakValueDictionary = WeakValueDictionary()
     filepath: Optional[Path] = None
 
+    def __new__(cls, filepath: Optional[Path] = None, *args, **kwargs):
+        """Creates a new model.
+        If the file at the provided file path was already parsed, this instance is returned.
+
+        Args:
+            filepath (Optional[Path], optional): The absolute file path to the file. Defaults to None.
+
+        Returns:
+            FileModel: A file model.
+        """
+
+        if filepath:
+            filepath = Path(filepath)
+
+            if filepath in FileModel._file_models_cache:
+                logger.info(
+                    f"Returning existing FileModel from cache, because {filepath} was already parsed"
+                )
+                return FileModel._file_models_cache[filepath]
+
+        return super().__new__(cls)
+
     def __init__(self, filepath: Optional[Path] = None, *args, **kwargs):
         """Initialize a model.
 
         The model is empty (with defaults) if no `filepath` is given,
         otherwise the file at `filepath` will be parsed."""
-        # Parse the file if path is given
+
         context_dir_reset_token = None
         if filepath:
             filepath = Path(filepath)  # so we also accept strings
 
-            # If a context is set, use it
-            if (folder := context_dir.get(None)) and not filepath.is_absolute():
-                logger.info(f"Used context to get {folder} for {filepath}")
-                filepath = folder / filepath
-            # Otherwise we're the root filepath
-            # and should set the context
-            else:
+            if not context_dir.get(None):
                 logger.info(f"Set context to {filepath.parent}")
                 context_dir_reset_token = context_dir.set(filepath.parent)
 
+            if filepath in FileModel._file_models_cache:
+                return
```

Comment:
If you return here (return None is more explicit), context_dir is never reset right?